### PR TITLE
Throw error on jsonDecode

### PR DESCRIPTION
**Pull request by Johann-S**
Throw an exception on jsonDecode method in ``Receiver.php``

Maybe it would be usefull to create custom Exception class for components like ``ComponentException`` any feedbacks ?

X-Ref : #35

### CODE REVIEW

**Comment by renatomefidf** [comment]
Hello @Johann-S , its indeed a good idea.

Check how JMS serializer does it:
https://github.com/schmittjoh/serializer/blob/master/src/JMS/Serializer/JsonSerializationVisitor.php#L27

In complement to that I think you can create a custom Exception which extends from \RuntimeException.
In case it is runtime since the json can be dynamic.

@gabrielrcouto and @reisraff what do you think about it?

**Comment by reisraff** [comment]
I think a good idea, but just remembering, as this exception will be an extended of \RuntimeException the \Gui\Application MUST catch this, and treat this exception, stopping the application or doing something that don't break the Application. And if the case be necessary to stop the application, should be good throw a message for the GUI explaining .why the application was finished.

**Comment by gabrielrcouto** [comment]
I agree with @reisraff, the exception can't stop the Application. We can develop an Alert component and use it for this. 

What you think guys?

**Comment by Johann-S** [comment]
I think it's a good way to display errors and handle exceptions :+1:

**Comment by reisraff** [comment]
#61

**Comment by Johann-S** [comment]
I added a ``ComponentException`` class and a comment to handle ``ComponentException`` in ``Receiver.php``.

I'll add a test when https://github.com/gabrielrcouto/php-gui/pull/63 is merged

File: src/Ipc/Receiver.php [hunk 1]
```diff
@@ -185,6 +186,7 @@ public function onData($data)
 
             if ($openingBraces > 0 && $openingBraces == $closingBraces) {
                 $messageJson = substr($this->buffer, $firstOpeningBracePos, $currentPos);
+                // TODO: Handle ComponentException
```
Comment (reisraff) [inline src/Ipc/Receiver.php]: is missing `@` for the docblock "@todo"

**Comment by Johann-S** [comment]
Done @reisraff

**Comment by gabrielrcouto** [comment]
Thank you @Johann-S, I liked the ComponentException :smile: 
#63 merged ;-)

**Comment by Johann-S** [comment]
Thank you @gabrielrcouto :smile: 

I need help, I'm not sure how can I test ``jsonDecode`` method because it's a ``protected`` method, any feedbacks ?

I wanted to use ``@expectedException ComponentException`` but my test failed

**Comment by reisraff** [comment]
I don't know if is a good idea but you test:

https://jtreminio.com/2013/03/unit-testing-tutorial-part-3-testing-protected-private-methods-coverage-reports-and-crap/

what do you think @gabrielrcouto

File: test/Ipc/ReceiverTest.php [hunk 2]
```diff
@@ -66,4 +67,36 @@ function ($result) use (&$foo) {
...
+    /**
+     * Call protected/private method of a class.
+     *
+     * @param object &$object    Instantiated object that we will run method on.
+     * @param string $methodName Method name to call
+     * @param array  $parameters Array of parameters to pass into method.
+     *
+     * @return mixed Method return.
+     */
+    public function invokeMethod(&$object, $methodName, array $parameters = array())
```
Comment (Johann-S) [inline test/Ipc/ReceiverTest.php]: I think this method should be moved to a Util PHPUnit class to be available for the others test class

**Comment by Johann-S** [comment]
All tests passed 

/CC @reisraff @gabrielrcouto

**Comment by gabrielrcouto** [comment]
Well done @Johann-S :smile:

**Pull request merged**